New submodules: ampcombi parsetables/complete/cluster

Darcy220606 opened the pull request:
## PR checklist

For this PR to PASS it needs approval of the new minimal test datasets #1189

This PR addresses the issue #4999 and closes PRs #4659 and #4533

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

Darcy220606 commented:
Note: For all the nf-tests, I asserted for the main process either the names of the folders/files because they all depend on a downloaded database that change the md5 sums.

Darcy220606 commented:
Hey @louperelo , do you know which pandas version was used i the old version of ampcombi ? The thing is its running locally but in github actions it seems like its running a new version.

pinin4fjords reviewed (changes requested):
First pass review here. A few common themes:

- Checking file names is a bit of a risky way to validate a process has run. It can be OK for a small subset of outputs, but if you do it with everything then you're not checking very hard that things have run well, you're not doing much more than what a stub run would do. Snapshot file content wherever possible (sometimes there's a random seed you can set for reproducibility), but failing that try to identify some diagnostic strings from the content and check as suggested in the nf-test docs I linked to.
- You can't just delete a module- people could be using it! You can deprecate though- I've provided a link.
- Make sure you indent for reproducibility.
- Order the meta.ymls to match the main.nfs.
- Use meta maps on the file inputs and outputs wherever possible.

Happy to come back and take another look after revisions.

jasmezz commented on modules/nf-core/ampcombi/complete/tests/main.nf.test:
```diff
@@ -0,0 +1,106 @@
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    file(process.out.tsv[0]).name,
```
Check for file contents if possible?

jasmezz commented on modules/nf-core/ampcombi/parsetables/meta.yml:
```diff
@@ -0,0 +1,107 @@
+  - DRAMP
+tools:
+  - ampcombi/parsetables:
+      description: "A parsing tool to convert and summarise the outputs from multiple AMP detection tools in a standradzed format."
```
```suggestion
      description: "A parsing tool to convert and summarise the outputs from multiple AMP detection tools in a standardized format."
```

jasmezz commented on modules/nf-core/ampcombi/parsetables/meta.yml:
```diff
@@ -0,0 +1,107 @@
+        e.g. `[ id:'sample1', single_end:false ]`
+  - amp_input:
+      type: list
+      description: The path to the directory containing the results for the AMP tools for each sample processed or a list of files corresponding to each file generated by AMP tools.
```
```suggestion
      description: The path to the directory containing the results for the AMP tools for each processed sample or a list of files corresponding to each file generated by AMP tools.
```

jasmezz commented on modules/nf-core/ampcombi/parsetables/meta.yml:
```diff
@@ -0,0 +1,107 @@
+      pattern: "*.faa"
+  - gbk_input:
+      type: file
+      description: The path to the file corresponding to the respective annotated files with either '.gbk' or '.gbff' extensions. Filenames must contain the corresponding sample-name, i.e. sample_1.faa
```
```suggestion
      description: The path to the file corresponding to the respective annotated files with either '.gbk' or '.gbff' extensions. File names must contain the corresponding sample name, i.e. sample_1.faa where "sample_1" is the sample name.
```

jasmezz commented on modules/nf-core/ampcombi/parsetables/tests/main.nf.test:
```diff
@@ -0,0 +1,75 @@
+        when {
+            process {
+                """
+                amp_input = [[id:'sample_1'],
+                             [file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/ampir/sample_1/sample_1.ampir.tsv', checkIfExists: true),
+                              file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/amplify/sample_1/sample_1.amplify.tsv', checkIfExists: true)]
+                              ]
```
```suggestion
                amp_input = [
                    [id:'sample_1'],
                    [
                        file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/ampir/sample_1/sample_1.ampir.tsv', checkIfExists: true),
                        file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/amplify/sample_1/sample_1.amplify.tsv', checkIfExists: true)
                    ]
                ]
```

jasmezz commented on modules/nf-core/ampcombi/parsetables/tests/main.nf.test:
```diff
@@ -0,0 +1,75 @@
+        when {
+            process {
+                """
+                amp_input = [[id:'sample_1'],
+                             [file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/ampir/sample_1/sample_1.ampir.tsv', checkIfExists: true),
+                              file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/amplify/sample_1/sample_1.amplify.tsv', checkIfExists: true)]
+                              ]
```
```suggestion
                amp_input = [
                    [id:'sample_1'],
                    [
                        file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/ampir/sample_1/sample_1.ampir.tsv', checkIfExists: true),
                        file('https://github.com/nf-core/test-datasets/raw/modules/data/delete_me/ampcombi/test_files_0.2/amplify/sample_1/sample_1.amplify.tsv', checkIfExists: true)
                    ]
                ]
```

jasmezz reviewed (changes requested):
What @pinin4fjords says and (as you know from me), many minor typo/alignment style suggestions. :grin:

Darcy220606 commented on modules/nf-core/ampcombi/main.nf:
```diff
@@ -1,52 +0,0 @@
-process AMPCOMBI {
```
When i add the deprecation the nf-tests dont pass through, and so what i did was i added the nf-tests for ampcombi (the old version), ran the nf-tests successfully and then added the deprecation message. But im afraid this old version wont pass the github action tests. 

Also to leave the old version of ampcombi, i had to rename the new one under another directory name `ampcombifilter` to differentiate between them to run the nf-tests. Is that ok you think ?

Darcy220606 commented on modules/nf-core/ampcombi/cluster/main.nf:
```diff
@@ -0,0 +1,48 @@
+        'biocontainers/ampcombi:0.2.2--pyhdfd78af_0' }"
+
+    input:
+    path(summary_file)
```
We dont need to identify the sample ID in both `ampcombi/complete` and `ampcombi/cluster`, because in teh original module it was devloped that it takes a list of files (the result from `ampcombi/parsetables`) and so when implemented in a pipeline all the files will be mixed in a channel and then input into the submodules as a list that doesnt take into account the sample ID.

Darcy220606 commented on modules/nf-core/ampcombi/cluster/meta.yml:
```diff
@@ -0,0 +1,48 @@
+      pattern: "*.tsv"
+
+output:
+  - versions:
```
done

Darcy220606 commented on modules/nf-core/ampcombi/cluster/tests/main.nf.test:
```diff
@@ -0,0 +1,109 @@
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    file(process.out.cluster_tsv[0]).name,
```
Checked the content now in all submodules

Darcy220606 commented on modules/nf-core/ampcombi/cluster/tests/nextflow.config:
```diff
@@ -0,0 +1,40 @@
+
+    withName: AMPCOMBI_PARSETABLES_1 {
+
+    ext.args = [
```
this is now not required as i dont need the setup processes anymore.

Darcy220606 commented on modules/nf-core/ampcombi/complete/main.nf:
```diff
@@ -0,0 +1,43 @@
+        'biocontainers/ampcombi:0.2.2--pyhdfd78af_0' }"
+
+    input:
+    path(summaries)
```
These two submodules `ampcombi/complete`and `ampcombi/cluster` are not meant to be used more than once in a workflow. They are summarizing tools that take the channels from all 'mixed' channels and merges them together. Therefore, no meta is required.

Darcy220606 commented on modules/nf-core/ampcombi/cluster/tests/main.nf.test:
```diff
@@ -0,0 +1,109 @@
+
+    config "./nextflow.config"
+    setup {
+        run("AMPCOMBI_PARSETABLES", alias:"AMPCOMBI_PARSETABLES_1") {
```
Ok so i have now added minimal test datasets for `ampcombi/complete` in a PR https://github.com/nf-core/test-datasets/pull/1189. And then i use `ampcombi/complete` output as a setup for `ampcombi/cluster`.

Darcy220606 commented on modules/nf-core/ampcombi/complete/tests/nextflow.config:
```diff
@@ -0,0 +1,40 @@
+
+    withName: AMPCOMBI_PARSETABLES_1 {
+
+    ext.args = [
```
This is now removed as the `ampcombi/parsetables` are not in the setup anymore.

Darcy220606 commented on modules/nf-core/ampcombi/ampcombi/tests/main.nf.test:
```diff
@@ -0,0 +1,70 @@
+                                process.out.summary_csv.collect { file(it[1]).getName() } +
+                                process.out.summary_html.collect { file(it[1]).getName() } +
+                                process.out.log.collect { file(it[1]).getName() } +
+                                process.out.results_db.collect { file(it[1]).getName() } +
```
Now added chacks for complete line numbers or strings.

jasmezz commented on modules/nf-core/ampcombi2/parsetables/tests/nextflow.config:
```diff
@@ -0,0 +1,22 @@
+    withName: AMPCOMBI2_PARSETABLES {
+
+    ext.args = [
+
```
```suggestion
```

jasmezz reviewed (approved):
LGTM, only tiny formatting stuffs. :smiley_cat:

Darcy220606 commented on modules/nf-core/ampcombi2/complete/tests/main.nf.test:
```diff
@@ -0,0 +1,56 @@
+    tag "ampcombi2"
+    tag "ampcombi2/complete"
+
+    test("ampcombi2_complete - bacteroides_fragilis") {
```
well spotted

SPPearce commented:
Is there anything outstanding to merge this in @Darcy220606 ?

Darcy220606 commented:
> Is there anything outstanding to merge this in @Darcy220606 ?

@SPPearce Ive addressed all the suggestions already, so it should be ready to merge once dev is merged in.
